### PR TITLE
ddl: support modify decimal column precision in some cases.

### PR DESCRIPTION
**Pull request by crazycs520**
### background

PR https://github.com/pingcap/tidb/pull/10433 prohibit modify decimal column precision, because change decimal column precision may cause data inconsistency when there is an index covering the decimal column.

### What problem does this PR solve? 

After think over, we can support modify a decimal column when satisfying the	following	conditions:
* No index covers the modified decimal column.
* Only support enlarges the decimal precision.
    * modify `decimal(5,2)` to `decimal(10,5)` is ok.
    * modify `decimal(5,2)` to `decimal(4, 2)` is not support.
    * modify `decimal(5,2)` to `decimal(5, 3)` is not support, because it may cause the column out range value error. 

**After TiDB supports all column type changes, there will be no such restrictions.**

eg:

```sql
>create table t (a decimal(5,2),index(a));
Query OK
>alter table t change a a decimal(4,2);
(8200, u"Unsupported modify column: can't change decimal column precision with index covered now")
>alter table t drop index a;
Query OK
>alter table t change a a decimal(4,2);
(8200, u'Unsupported modify column: length 4 is less than origin 5')
>alter table t change a a decimal(5,3);
(8200, u'Unsupported modify column: modify original decimal(5,2) to decimal(5,3) may cause out of range value error')
>alter table t change a a decimal(10,5);
Query OK
```
### What is changed and how it works?

Below is an example if only change `ddl` pkg to support `modify decimal precision`  : 

1. The original decimal precision is `decimal(2,1)`, and write the data `1.1`
2. modify the decimal precision to `decimal(4,2)`, and get the data `1.1`, **but the expect data is `1.10`.**

Below is sql example:

```sql
test> create table t (a decimal(2,1));
test> insert into t set a=1.1;
test> select * from t;
+-----+
| a   |
+-----+
| 1.1 |
+-----+
test> alter table t change a a decimal(4,2);
test> select * from t;
+-----+
| a   |
+-----+
| 1.1 |   -- the value should be 1.10
+-----+
```

Since the decimal already encodes the precision info, So after modifying the decimal column precision, **we must convert the decimal value to the new precision when reading the old decimal value**. 

So Also need to change the `decode` logical, **both in TiDB and TiKV**.

Related TiKV PR https://github.com/tikv/tikv/pull/6532.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Related changes

 - Need to cherry-pick to the release branch?  
 - Need to update the documentation

Release note

 - Support modify decimal column precision in the case like below:
    * No index covers the modified decimal column.
    * Only support enlarges the decimal precision.

### CODE REVIEW

**Comment by crazycs520** [comment]
@zimulala @bb7133 @coocood PTAL

**Comment by bb7133** [comment]
> Support modify decimal column precision in some cases.

Please update the release note by adding descriptions to the cases.

File: util/codec/codec.go [hunk 1]
```diff
@@ -989,7 +989,7 @@ func (decoder *Decoder) DecodeOne(b []byte, colIdx int, ft *types.FieldType) (re
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		if ft.Decimal != types.UnspecifiedLength && frac > ft.Decimal {
+		if ft.Decimal != types.UnspecifiedLength && frac != ft.Decimal {
```
Comment (zimulala) [inline util/codec/codec.go]: Why change this code? If we don't change it, we also can pass all the tests as well.

Comment (crazycs520) [reply util/codec/codec.go]: No, now, you will fail in the test `TestModifyColumnWithDecimal`

Comment (zimulala) [reply util/codec/codec.go]: It's my mistake. Ignore it.

File: util/codec/codec.go [hunk 2]
```diff
@@ -714,6 +714,28 @@ func DecodeRange(b []byte, size int) ([]types.Datum, []byte, error) {
 	return values, nil, nil
 }
 
+// DecodeOneWithTp decodes on datum from a byte slice generated with EncodeKey or EncodeValue.
+func DecodeOneWithTp(b []byte, ft *types.FieldType) (remain []byte, d types.Datum, err error) {
+	remain, d, err = DecodeOne(b)
+	if ft.Tp != mysql.TypeNewDecimal || err != nil {
+		return remain, d, err
+	}
+	if ft.Decimal != types.UnspecifiedLength && d.Frac() != ft.Decimal && d.GetInterface() != nil {
```
Comment (zimulala) [inline util/codec/codec.go]: Why not do it in `DecodeOne`?

Comment (crazycs520) [reply util/codec/codec.go:723]: Because `DecodeOne` doesn't know the `ft` parameter.

File: table/tables/index.go [hunk 3]
```diff
@@ -172,10 +175,18 @@ func TruncateIndexValuesIfNeeded(tblInfo *model.TableInfo, idxInfo *model.IndexI
 					v.SetString(v.GetString(), tblInfo.Columns[ic.Offset].Collate)
 				}
 			}
+		case types.KindMysqlDecimal:
```
Comment (wjhuang2016) [inline table/tables/index.go:178]: Why we need this?

**Pull request closed without merging**